plugins/BEdita/Core/src/Cache/LayeredEngine.php

```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2021 ChannelWeb Srl, Chialab Srl
+ *
+ * This file is part of BEdita: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+namespace BEdita\Core;
+
+use Cake\Cache\Cache;
+use Cake\Cache\CacheEngine;
+use Cake\Cache\Engine\ArrayEngine;
+use Exception;
+
+/**
+ * This engine uses two layers of cache, one persistent and one in-memory for faster lookup times.
+ */
+class LayeredEngine extends CacheEngine
+{
+    /**
+     * Persistent cache instance.
+     *
+     * @var CacheEngine
+     */
+    protected $_persistent = null;
+
+    /**
+     * In-memory cache instance.
+     *
+     * @var CacheEngine
+     */
+    protected $_memory = null;
+
+    /**
+     * The default config used unless overridden by runtime configuration
+     *
+     * - `persistent` A cache configuration or an alias, to use as persistent cache
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'persistent' => ['className' => 'File'],
+    ];
+
+    /**
+     * {@inheritDoc}
+     * @throws Exception If the configuration is wrong
+     */
+    public function init(array $config = []): bool
+    {
+        parent::init($config);
+
+        $this->_persistent = $this->getEngineInstance($this->getConfig('persistent'));
+        $this->_memory = new ArrayEngine();
+
+        return true;
+    }
+
+    /**
+     * Get the engine instance from a configuration or an alias.
+     *
+     * @param array|string $config Engine configuration or an alias
+     * @return CacheEngine The engine instance
+     * @throws Exception If the configuration is wrong
+     */
+    protected function getEngineInstance($config): CacheEngine
+    {
+        $registry = Cache::getRegistry();
+
+        if (is_string($config)) {
+            if (!$registry->has($config)) {
+                throw new Exception("Cache engine alias {$config} is not defined");
+            }
+
+            $instance = $registry->get($config);
+
+            if (!$instance instanceof CacheEngine) {
+                throw new Exception("Cache engine alias {$config} is not an implementation of CacheEngine");
+            }
```

Comment:
We could probably raise an exception if `$instance === $this`.

Imagine the scenario `'default' => ['className' => 'BEdita/Core.Layered', 'persistent' => 'default']`. This would end up in an infinite recursion when accessing the `$_persistent` engine.

Reply:
While writing the tests, I found out that a configuration like the one you suggested would be caught by the previous condition, because the engine itself is not yet instanced:
```
if (!$registry->has($config)) {
    throw new Exception("Cache engine alias {$config} is not defined");
}
```

Reply:
Ok, since this is an edge case, I'd say that we could leave the sanity check in place, but while testing let's just ensure it raises an exception, without caring too much about its class, message or code. A 100% coverage is a good thing, but if it is only achievable with workarounds, I think it's best to keep a line uncovered.

Doing so, we do test what we _actually_ care about — the fact that some sort of exception is raised if the passed configuration would result in an infinite loop — although the line where we raise the exception will remain uncovered, so it stays there just in case Cake changes the way it handles its registry or whatever other reason, and it acts as a reserve parachute.